src/Microsoft.AspNetCore.Razor.Evolution/CodeGeneration/RedirectedRuntimeBasicWriter.cs

```diff
@@ -79,8 +91,8 @@ public override void WriteHtmlContent(CSharpRenderingContext context, HtmlConten
                     .WriteStringLiteral(textToRender)
                     .WriteEndMethodInvocation();
 
-                charactersConsumed += textToRender.Length;
+                    charactersConsumed += textToRender.Length;
```

Comment:
You can rewrite this code to be much more efficient by going direct to `context.Writer` and looking at the lengths of the strings before you write them.

Also `context.Writer` should expose a way to write a portion of a string using index+length